05_testing_and_ci/boost_testing_sidemade_demo.md

```diff
@@ -0,0 +1,208 @@
+# Boost.Test and CTest in Action: SideMade Demo
+
+Repository: [testing boost exercise – demo-start branch](https://github.com/Simulation-Software-Engineering/testing-boost-exercise/tree/demo-start)
+
+## (1) Get to Know the Code
+
+- intro: go through all sections of `README.md`
+- try: build and run
+    - explain shell output
+    - try different matrices and algorithms
+- look into the code
+    - `main.cpp`
+    - `Configuration.hpp/cpp`
+    - `matrixIO.hpp/cpp`
+    - `MatrixSolver.hpp/cpp`
+- look into `CMakeLists.txt`
+- our mission
+    - The code has no unit tests yet. -> We will some of them with Boost UTF.
+    - exercise: continue with same code, add more tests, set up GitHub Actions, and more
+
+## (2) First Dummy Test Case
+
+- `mkdir tests`, create `tests/MatrixSolverTest.cpp`
+- `#define BOOST_TEST_MODULE SideMadeTests`
+    - needs to be in one of our test files, we will write multiple ones
+- `#define BOOST_TEST_DYN_LINK`
+    - needed if UTF uses as dynamic library
+    - in header-only mode, you cannot split tests over mutliple files
+    - needed in every file
+- `#include <boost/test/unit_test.hpp>`
+    - different header/path than for header-only usage
+- add first test
+
+```cpp
+BOOST_AUTO_TEST_CASE(LU)
+{
+  BOOST_TEST(true);
+}
+```
+
+- adjust `CMakeLists.txt`
+
+```cmake
+include(CTest)
+find_package(Boost 1.71 REQUIRED unit_test_framework)
+file(GLOB_RECURSE TEST_FILES CONFIGURE_DEPENDS tests/*.cpp)
+add_executable(testsidemade "${TEST_FILES}")
+target_link_libraries(testsidemade PRIVATE Boost::unit_test_framework)
+add_test(NAME "MatrixSolverTests" COMMAND ${CMAKE_CURRENT_BINARY_DIR}/testsidemade)
+```
+
+- reconfigure cmake, build, run via
+    - `./testsidemade`
+    - `./testsidemade --list_content`
+    - `make test`
+
+## (3) Use Actual Implementation in Matrix Solver Test
+
+- need more includes
+
+```cpp
+#include <Eigen/Dense>
+#include "MatrixSolver.hpp"
+
+using namespace Eigen;
+```
+
+- create data
+    - show how to manufacture a testcase (a common scheme in numerics)
+    - first A, then x, then b
+
+```cpp
+MatrixXd A(3, 3);
+A << 1, 2, 3,
+     4, 5, 6,
+     7, 8, 9;
+
+VectorXd b(3);
+b << 3.5, 11, 18.5;
+
+VectorXd expectedX(3);
+expectedX << 2, 0, 0.5;
+```
+
+- add testing code
+
+```cpp
+MatrixSolver solver(MatrixSolver::LU);
+VectorXd     x(3);
+solver.solve(A, b, x);
+BOOST_TEST(x == expectedX);
+```
+
+- adjust `CMakeLists.txt`
+    - add `"${SRC_FILES}"` to test executable
+    - makes headers visible `target_include_directories(testsidemade PRIVATE ${PROJECT_SOURCE_DIR}/src)`
+    - link `Eigen3::Eigen` and also `yaml-cpp` (since used in `SRC_FILES`)
+
+- build ... problem that two "main" functions are defined
+    - Fix by:
+
+```cmake
+file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS src/*.cpp)
+list(REMOVE_ITEM SRC_FILES "${PROJECT_SOURCE_DIR}/src/main.cpp")
```

Comment:
Maybe we will have a student that shows us an even better solution to remove `main.cpp` from the source files. :)